Support User: Bypass localForage while active to avoid conflicts across sessions

jordwest opened the pull request:
Bypasses localForage while support user is active. This allows sync-handler and other persistent data stores using localForage to use a 'sandboxed' temporary in-memory store while support user is active, avoiding conflicts with the original user's persisted data. This doesn't cover instances of direct localStorage usage - see #3941 which covers that.

The goal is to eventually move all direct localStorage usage to localForage.

-----

This still needs some more testing, but overall it's working.

`localforage-bypass.js` was copied and modified from a [localForage testing module](https://github.com/mozilla/localForage/blob/master/test/dummyStorageDriver.js); ~~we can likely cut quite a bit of cruft out of it as a lot of the logic relates to test facilitation.~~ (cruft removed in 4c39e79)

Things to test:
 - [x] With `support-user` feature flag disabled
 - [x] With `wpcom_user_bootstrap` enabled
 - [x] Running in non-`development` environment
 - [x] Running in `development`
 - [x] Chrome 49
 - [x] Firefox
 - [x] Safari
 - [x] IE


cc @retrofox @gwwar

jordwest commented:
@gwwar After a lot of trial and error, what we chatted about earlier worked :smiley:  Would you mind taking a look at this solution? Basically to solve the issues with the fact that `localForage.defineDriver` is async:

 - **The `build-env.*.js` and other Calypso specific bundles are not loaded on boot**. Instead we make sure support user has done any pre-boot async tasks (in this case running `localForage.defineDriver`) before loading the Calypso bundle. This only happens when the query string is set:
 - **A `?support_user=true` query string is passed to inform the server that this will be a support user session.** This allows the server to leave out any user specific info including the user bootstrap and to inject a special pre-boot sequence.
 - **`localforage` package was added to the `vendor` bundle** so that the same instance is shared by both the support user pre-boot bundle and the main Calypso bundles.

After this change, it should even be possible to remove some of the changes introduced in #3879, as the Redux store would no longer need to know about the support user state. Instead, because `initialReduxState` is undefined and any attempts to save/load from localForage will be bypassed, the Redux store should function as expected without being aware that it's sandboxed.

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+// original user/vice versa.
+// Copied with modifications from localForage source: localforage/test/dummyStorageDriver.js
+
+var dummyStorage = {};
```
const

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+function _initStorage( options ) {
+	const self = this;
+
+	let dbInfo = {};
```
This can be a const. We don't reassign the dbInfo reference.

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+		}
+	}
+
+	dummyStorage[ dbInfo.name ] = dbInfo.db = {};
```
I'm personally not a huge fan of setting multiple variables in a single line.

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+
+	let dbInfo = {};
+	if ( options ) {
+		for ( let i in options ) {
```
Another way:
```javascript
options.forEach( ( value, index ) => {
    dbInfo[ index ] = value;
} );
```

gwwar commented on client/lib/user/support-user-interop.js:
```diff
@@ -82,7 +87,7 @@ export const rebootWithToken = ( user, token ) => {
 
 	store.set( STORAGE_KEY, { user, token } );
-	window.location.reload();
+	window.location.search = 'support_user=true';
```
Note for anyone reading, that this also triggers a refresh.

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+
+// Config the localStorage backend, using options set in the config.
+function _initStorage( options ) {
+	const self = this;
```
Are we sure we need `self`?

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+	debug( 'localForage bypass', 'clear' );
+
+	const self = this;
+	const promise = new Promise( function( resolve, reject ) {
```
Two other options would be to use `.bind` or fat arrow, to avoid creating the self const:
```javascript
const promise = new Promise( function( resolve, reject ) => {
//...
} ).bind( this ) );
```
```javascript
const promise = new Promise( ( resolve, reject )  => {
```

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+				for ( let key in db ) {
+					const result = db[ key ];
+
+					callback( result, key );
```
Does this actually iterate on following calls? Unit tests are highly suggested. :)

This might be a fun place to play with `yield`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/yield

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+			let index = 0;
+
+			for ( let key in db ) {
+				if ( db.hasOwnProperty( key ) && db[ key ] !== undefined ) {
```
Too bad `Object.keys` doesn't preserve order

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+	const promise = new Promise( function( resolve, reject ) {
+		self.ready().then( function() {
+			const db = self._dbInfo.db;
+			let keys = [ ];
```
Again, `const` here :smile: 

> The const declaration creates a read-only reference to a value. It does not mean the value it holds is immutable, just that the variable identifier cannot be reassigned.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const

gwwar commented on client/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,242 @@
+
+	// Cast the key to a string, as that's all we can set as a key.
+	if ( typeof key !== 'string' ) {
+		key = String( key );
```
Have we ever run into this case? I think it would be ok to reject here too.

gwwar commented on server/pages/index.jade:
```diff
@@ -110,25 +110,45 @@ html(lang=lang, dir=isRTL ? 'rtl' : 'ltr', class=isFluidWidth ? 'is-fluid-width'
 		if app
 			script(type='text/javascript')!='var app = ' + sanitize.jsonStringifyForHtml( app )
-		if initialReduxState
+		if initialReduxState && !supportUser
```
cc @seear, heads up for support user behavior.

seear replied:
`initialReduxState` here is not data from localStorage. It is API response data that has been used to initially render the requested page on the server, and is here being passed to the client so that the client will not need to fetch data and re-render. It is only ever used for logged-out pages. I believe this behaviour needs to be consistent whether the support user is active or not, so we probably don't want this check added here.

Thanks for the ping @gwwar

jordwest replied:
If support user is active, the API data injected here will belong to the original user and not the supported user, which we'd like to ignore while support user is active. We'll need to fetch the data and re-render after the support user token has been applied anyway, which doesn't happen until later in the Calypso boot process. Could we instead initialize it with something like `initialReduxState = {}` when support user is active?

seear replied:
Server rendering and subsequent passing of the data used for that render to the client in `initialReduxState` will only ever happen when no user is logged in, and there are no plans to do this for logged-in users, so I think this discussion is academic.

If we did theoretically want to make sure to not use original user data at all, we'd have to do so at an earlier point, to avoid the server rendering with that data. For now I would simply remove this check, since `initialReduxState` would never be populated when supportUser is active (assuming supportUser only affects logged-in operation).

mtias replied:
Yes, I'm curious why we need `supportUser` checks on server side pages at all. Seems to add quite a bit of complexity.

seear replied:
For reference [`server-side-rendering`](https://github.com/Automattic/wp-calypso/blob/f9f0d81e41d9d3f8d33e72772b52ff5239a112af/config/wpcalypso.json#L85) is the relevant feature flag, if we ever want to make sure server rendering does not occur when support user is active.

jordwest replied:
@seear: Ah, thanks for the insight. That makes sense - I'll move the logic into the initial context setup.

@mtias: We need to make sure that no original user information is left in the page, or we'll end up with inconsistent data - some that belongs to the original user, some that belongs to the supported user. There is unfortunately a fair bit of complexity introduced by this feature; it's like adding multi-user support to single-user software :smile: I think moving these checks to the server side context setup should tidy this up though.

Let's move this discussion to #4068 as it explains the reasoning more clearly and in detail without the localForage changes.

gwwar commented:
:+1: Nice approach. I'll clean up #3879 after this lands. I verified that we write to a dummy driver when we switch users. 

We may want some more :eyes: for the bundle updates. Maybe @mtias or @aduth?

jordwest commented:
Thanks for the review @gwwar. I've cleaned up `localforage-bypass.js` and added unit tests. The PR is getting quite large now, so I've split out the new pre-boot sequence to a separate PR: #4068. Once that lands this one should reduce in size.

aduth commented on client/lib/support/support-user/test/localforage-bypass.js:
```diff
@@ -0,0 +1,185 @@
+	describe( 'length', () => {
+		const length = () => localForage.length();
+		const expectLength = ( expected ) => {
+			return () => {
```
Does this need to be nested?

aduth commented on client/lib/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,230 @@
+
+export default {
+	_driver: 'localForageBypass',
+	_initStorage: _initStorage,
```
You could alternative declare them as exports at the time of being declared. Not a huge advantage, but might help with clarity of what's being exported.

```js
export const _driver = 'localForageBypass';

export function _initStorage( options ) {
	// ...
}
```

aduth commented on client/lib/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,230 @@
+	const dbInfo = {};
+	if ( options ) {
+		for ( let i in options ) {
+			dbInfo[ i ] = options[ i ];
```
Seems these handful of lines be replaced with a single [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)?

```js
const dbInfo = Object.assign( {}, options, { db: {} } );
```

aduth commented on client/lib/support/support-user/test/localforage-bypass.js:
```diff
@@ -0,0 +1,185 @@
+		};
+
+		it( 'should be zero when initialized', ( done ) => {
+			expectLength( 0 )()
```
Mocha has built-in support for promises, so you don't need to call `done` yourself:

https://mochajs.org/#working-with-promises

jordwest replied:
Ah cool, I had used chai-as-promised but wasn't aware similar functionality is built in to mocha, thanks!

aduth commented on client/lib/support/support-user/support-user-interop.js:
```diff
@@ -82,7 +89,9 @@ export const rebootWithToken = ( user, token ) => {
-	window.location.reload();
+
+	// This triggers a reboot
+	window.location.search = 'support_user=true';
```
Do we care to preserve existing search parameters?

jordwest replied:
Not particularly, as we'll be switching to another user anyway so any search context won't be relevant. Potentially we might even want to default to sending the user to `/me`.

aduth commented on client/lib/support/support-user/pre-boot.js:
```diff
@@ -0,0 +1,46 @@
+		debug( 'Booting Calypso bundle', window.supportUserReady );
+
+		loadScript( script, () => {
+			loadedScripts++;
```
Seems like something [`Promise.all`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) is particularly well-suited for: 

```js
const loadScript = ( script ) => {
	return new Promise( ( resolve ) => {
		// ...
		bundle.onload = resolve;
	} );
};

Promise.all( readyScripts.map( ( script ) => loadScript( script ) ).then( window.AppBoot );
```

aduth commented on client/lib/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,230 @@
+// in-memory store which is lost on page reload. This driver is used to sandbox
+// a user's data while support-user is active, ensuring it does not contaminate the
+// original user/vice versa.
+// Copied with modifications from localForage source: localforage/test/dummyStorageDriver.js
```
A small thing, but we may want to point to [`CREDITS.md`](https://github.com/Automattic/wp-calypso/blob/master/CREDITS.md#httpsgithubcommozillalocalforage) for licensing information.

aduth commented on client/lib/support/support-user/localforage-bypass.js:
```diff
@@ -0,0 +1,230 @@
+export default {
+	_driver: 'localForageBypass',
+	_initStorage: _initStorage,
+	// _supports: function() { return true; }
```
Drop commented code.

aduth commented:
I feel we're starting to have quite a few different pre-boot behaviors, each with their own unique take on how to initialize prior to the "real boot". Is there any particular reason we couldn't have used similar logic to that of the `render-visualizer` dev modules or Redux store initialization here as well? I might even imagine something more generalized:

```js
function boot() {
	// ...
}

const preBoot = [ devTools, supportUser, createReduxStore ];
Promise.all( preBoot ).then( boot );
```

jordwest commented:
Thanks for the review @aduth.

> Is there any particular reason we couldn't have used similar logic to that of the render-visualizer dev modules or Redux store initialization here as well?

That has worked well up until now, as long as the main boot sequence is synchronous. The challenge here is that there are actually multiple boot sequences happening asynchronously, due to modules doing their init as soon as they're `require`d. You can verify this by commenting out `window.AppBoot()` in `server/pages/index.jade` and notice that many Calypso modules boot regardless. Here's a diagram to illustrate - the numbers show the execution order:

![async](https://cloud.githubusercontent.com/assets/416133/13831757/c889db1e-ec21-11e5-969d-307e40c3a10b.jpg)

By introducing async to the Calypso boot sequence, we would give the other modules an opportunity to do their initialization which includes accessing localForage. In this case, we don't want them accessing localForage until this bypass driver has been defined, otherwise they will be accessing real data in the time between (3) and (8) above.

The solution in #4068 is to not even load the webpack bundle until the async `localForage` booting is done:
![calypso boot - support user](https://cloud.githubusercontent.com/assets/416133/13831848/7299d78a-ec22-11e5-98b6-0ec116793a2a.jpg)

Although it seems excessive, this extra step will only be done when support user is enabled; this method seemed a good compromise as it would have no effect on those that aren't using this feature.

jordwest commented:
Marked this PR as "In Progress" until #4068 lands or we can come up with an alternative.

jordwest commented:
This now uses the changes that landed in #4114 to bypass localForage on startup. Here's an overview of how we overcame the async issue described above:

![localforage_bypass](https://cloud.githubusercontent.com/assets/416133/13976627/511be52c-f10d-11e5-8c95-0eff3cfd2740.jpg)

gwwar commented on client/lib/localforage/index.js:
```diff
@@ -16,11 +21,15 @@ const config = {
-} );
+const localForagePromise = localforage.defineDriver( localforageBypass )
+	.then( () => {
+		localforage.config( config );
```
What does the timing look like if we check if we're in switch user mode here? Unfortunately I'm still seeing the sync-handler write to indexedDB while switched.

jordwest replied:
Hmm, that's weird, are you seeing a `Cannot bypass localforage after initialization` error in the console?

jordwest replied:
Oh :smile:

gwwar commented:
![giphy](https://cloud.githubusercontent.com/assets/1270189/14004012/54f79cfc-f114-11e5-8a6a-1b60af2af483.gif)

jordwest commented:
![hi5](https://cloud.githubusercontent.com/assets/416133/14004934/d1ce3cdc-f1a9-11e5-83cd-0bc8cf8c7770.gif)
